torch/distributed/_shard/sharded_tensor/api.py

```diff
@@ -273,46 +288,76 @@ def gather(
                 Must to be provided ONLY on ``dst`` rank.
                 Default: ``None``
         """
+        def shard_size(shard_md):
+            return math.prod(shard_md.shard_sizes)
+
         rank = dist.get_rank(self._process_group)
         full_size = self.metadata().size
         _validate_output_tensor_for_gather(rank, dst, full_size, out)
 
         local_shards = self.local_shards()
 
         world_size = dist.get_world_size(self._process_group)
-
-        gathered_shards: List[Optional[List[Shard]]] = [None] * world_size if rank == dst else []
-        # TODO: see how we could use dist.gather() instead of dist.gather_object
-        # as the latter one involves pickling on CPU, see more context
-        # https://github.com/pytorch/pytorch/issues/73935
-        dist.gather_object(
-            obj=local_shards,
-            object_gather_list=gathered_shards,
+        rank_sizes = [0 for _ in range(world_size)]
+        max_rank_size = 0
+        shard_placement: Dict[int, Tuple[int, int]] = dict()
+        local_shards_offset = []
+        # collect sizes
+        for shard_idx, shard_md in enumerate(self.metadata().shards_metadata):
+            shard_rank = cast(_remote_device, shard_md.placement).rank()
+            assert shard_rank is not None
+            shard_placement[shard_idx] = (shard_rank, rank_sizes[shard_rank])
+            if shard_rank == rank:
+                local_shards_offset.append((shard_md, rank_sizes[shard_rank],))
+
+            rank_sizes[shard_rank] += shard_size(shard_md)
+            max_rank_size = max(max_rank_size, rank_sizes[shard_rank])
+
+        gather_list: Optional[List[torch.Tensor]]
+        if rank == dst:
+            assert out is not None
+            gather_list = [torch.empty((max_rank_size,), device=out.device) for _ in range(world_size)]
+        else:
+            gather_list = None
+
+        with torch.no_grad():
+            data = torch.empty(max_rank_size, device=self._get_preferred_device())
+            for shard in local_shards:
+                for placement in local_shards_offset:
```

Comment:
nit: Might be cleaner if `local_shards_offsets` is a dict keyed on `ShardMetadata`, since `ShardMetadata` is a dataclass it probably already has `__eq__` and `__hash__` to enable this, if not maybe we can add it?

Reply:
I had to make ShardMetadata and _remote_device hashable.
Which those changes, the `gather` code became significantly simpler.
Good catch!

Reply:
I made that change, had to make both ShardMetadata and _remote_device  hashable.
The code turned out to be significantly clearer.